### DISCUSSION TITLE
Changes needed to allow blocked off areas

### DISCUSSION THREAD

**Discussion by SilPho** [CorsixTH/CorsixTH#2214] [Research Room]
I've been playing around with allowing blocked-off areas in the hospital, and I noticed a few fairly obvious issues as it stands, which is obviously to be expected. So, I had a play around at fixing some of the worst problems that I found.

Why am I saying all of this? I'm curious to see if I'm at least on the right lines with these changes, and to find out if it's worth creating a PR for this early prototype.

Disclaimer: I'm an experienced web programmer, but this is my first time in Lua or any serious game dev, so I apologise if I say something stupid or naive.

**It's possible to create rooms with unreachable objects - Mostly fixed**
I changed the edit room dialogue to run a check that every item in the room (except for bins, radiators, and extinguishers) can be reached from the room's door `edit_room.lua, checkEnableConfirm()` This mostly prevents players from creating impossible rooms, but I've still had problems with the Operating Table (so far) since multiple tiles need to be checked for access. The other problem is that there's no feedback for the player to say why the button remains disabled.

**It's possible to place a staff member in an unreachable spot within a room - Fixed**
The game already prevents you from placing staff on top of other objects as you move the cursor around `place_staff.lua, onCursorWorldPositionChange(x, y)` so I added a check to make sure that cursor location can be pathed towards the room's door. This seems to work very well.

With just those two changes the game is very stable with the "Allows blocked off areas" flag enabled. In regular play I've encountered significantly fewer crashes. However, I am actively trying to avoid two major problems.

**It's possible to move furniture objects into a room later that break the flow of a room - Still broken**
This isn't much of a problem if you're careful, but while purchasing corridor furniture you can place things like plants into rooms, which could easily be placed in a way to block parts of the room. I'm sure it's probably possible to re-use the same "can-every-object-be-reached-from-the-door" logic, but I don't know how to do this (yet).

**It's possible to get staff into unreachable parts of the hospital - Very broken**
Getting staff into these blocked off areas doesn't break anything until they try to path into a room. You can get into this state by placing staff into the blocked off areas, or by blocking the areas off while staff are already there. The immediate cause of the crash here appears to be a text rendering bug with the Error message box?

**Comment by TheCycoONE** [comment]
Great, thank you for looking at this.

On item 1 I'm interested in the originals handling of rooms. If someone had time I'd be curious what the game does when you try to make a plant wall blocking the chair in the GP office for instance. That should give us some guidance.

Your last point was the primary motivation for the meta issue. It's worth reporting any of the traces you do get. The original had an Easter egg involving stuck staff - and an epidemic trick of blocking all the exits which currently cannot be replicated.

**Comment by SilPho** [comment]
The original game prevents a plant wall from dividing a room in the same way that CTH does - The object is greyed out when you try to place it. I just tested it on the Origin version.

The original does let you create blocked off areas in a room, as long as nothing vital is in the blocked off area.  For example, you can create a wall of plants like the image below, but not if the chair or filing cabinet is inside the wall, nor can you put one of those inside the wall afterwards either.
![Plants](https://user-images.githubusercontent.com/22282491/184978323-891e85e6-3d66-419d-8b55-2df930a936c9.jpg)
This behaviour is different to CTH's default behaviour, which won't allow any blocked tiles.

Here's an example of a trace when a doctor has been positioned in a blocked off area and then gets a call to enter a room. Specifically, a patient arrived at reception and the doctor was called to the GP's office.
[EmptyQueue.txt](https://github.com/CorsixTH/CorsixTH/files/9354421/EmptyQueue.txt)

**Reply by TheCycoONE** [reply]
Excellent, thank you!

**Comment by ARGAMX** [comment]
@SilPho sounds nice. 
I have noticed these problems with blocked off areas too. 
In Original TH blocked off areas enabled, but working much more stable, than in CTH.

How about a PR?
